### PR TITLE
MDN CSS: Add support for -webkit and -moz extensions

### PR DESCRIPTION
**Pull request by rasikapohankar**
Attempted to add fetching of URLs for the -webkit and -moz extensions in fetch.pl. For #293 
Please review :)

###### People to notify 

@moollaza @hchienjo 

IA Page: https://duck.co/ia/view/mdn_css

Maintainer: @hchienjo

### CODE REVIEW

**Comment by daxtheduck** [comment]
@rasikapohankarThanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by hchienjo** [comment]
@rasikapohankar Unfortunately it looks like you rearranged most of the code in the script quite a bit. Since the new urls share the same structure, the only change required was a bit of refactoring in [the fetching](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/371).

There is still more work to be done on this and I can help you help me make this.

**Comment by rasikapohankar** [comment]
Hi @hchienjo! Yes I did refactor some of the code, I put the part of the code that fetched into a subroutine called `fetch()` and called it for each of the URLs from the loop [here](https://github.com/rasikapohankar/zeroclickinfo-fathead/blob/6ca3f84fcee1183cd7629c506bae47508c85ecd3/lib/fathead/mdn_css/fetch.pl#L48) and the downloading part into `download()`, but I didn't change any of the earlier code.

Except I added [this](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/363/files#diff-ca443e93fd9afead3b245f141e784161R82) because the Mozilla extensions page had a similar `li` structure without hyperlinks.
I have also called `download()` for each of the URLs, but after looking at the code in your PR I realised it can be called only once!

Yes, I'd like some help. What should I do with this PR? Also is it a convention to not use subroutines in scripts?

**Comment by hchienjo** [comment]
@rasikapohankar No, it is not a convention to not use subroutines, you can use them as much as possible. Generally you should not have too many changes in one commit. 

@moollaza Help.

File: lib/fathead/mdn_css/fetch.pl [hunk 1]
```diff
@@ -17,114 +17,131 @@
...
+
+    if ( $tx->success ) {
+      my $divs = $tx->res->dom->find('div.index, div.column-half');
+        for my $div ( $divs->each ) {
+            for my $ul ( $div->find('ul')->each ) {
+                $ul->find('li')->map(
+                    sub {
+                        my $li = shift;
+                        print "\n value of li $li";
+                        if($li->at('a')) {
```
Comment (hchienjo) [inline lib/fathead/mdn_css/fetch.pl]: @rasikapohankar Yes, you were on the right track here! I [noticed it too](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/371/files#diff-ca443e93fd9afead3b245f141e784161R41)

**Comment by rasikapohankar** [comment]
@hchienjo Ooh, yes. Sorry about that!

**Comment by moollaza** [comment]
> @moollaza Help.

@rasikapohankar @hchienjo I'm a little confused -- what do you need help with? Is the code working? Is it ready for Beta?

**Comment by rasikapohankar** [comment]
@moollaza #371 and this PR are for the same task. What should we do? Yes the code works, but the output.txt has some titles like `-moz-appearance -webkit-appearance`. So this does is not giving an answer for `-moz-appearance`. 
I changed the code to have subroutines for fetch and download. Is that okay?

**Comment by moollaza** [comment]
@rasikapohankar @hchienjo it looks like both your PR's have the same end result -- however both are missing all the webkit articles. I only see one or two new lines that look incorrect.

I'm OK with the new subroutines.

@rasikapohankar can you resolve the conflicts and investigate why there aren't any webkit results?

Once we have webkit we can do something like `s/\(-webkit.+\)//;` to better handle that first article. The others all look fine

**Comment by rasikapohankar** [comment]
@moollaza The conflict is due to output.txt, I had pushed it so it could be verified. Should I remove it? 
@moollaza Yes I will investigate the webkit entries but there are some on lines 78-95..

**Comment by rasikapohankar** [comment]
@moollaza Only some of the properties under `Proprietary WebKit-prefixed entities (do not use on Web sites)` on https://developer.mozilla.org/en-US/docs/Web/CSS/Webkit_Extensions are redirecting to links of the form `-webkit-x-..`

the others redirect to unprefixed properties, for example, `-webkit-column-count` link goes to https://developer.mozilla.org/en-US/docs/Web/CSS/column-count.

That's why only few -webkit entries are seen in the file?

Should these -webkit properties that link to unprefixed properties have entries as R `-webkit-property` in the output.txt?

**Comment by rasikapohankar** [comment]
@moollaza The conflicts are not present now, I reverted the commit of the output file.

Can it be deployed to beta?

**Comment by moollaza** [comment]
> Should these -webkit properties that link to unprefixed properties have entries as R -webkit-property in the output.txt?

Yes definitely it looks like @hchienjo has partially accomplished this in his PR now

**Comment by rasikapohankar** [comment]
That is great! @moollaza Can you please suggest what I should do here next? Or should this just be closed?

Also, for the -moz-appearance (-webkit-appearance) article we talked about, I tried and a regex like `m/(.+) (?:\((:.+|-webkit.+)\))/` will work, since we don't want `(.+) (.+)`. I'd like to update the parser if this looks okay.

**Comment by moollaza** [comment]
> Can you please suggest what I should do here next? Or should this just be closed?

@rasikapohankar I'm deploying this to Beta right now. So far it looks great 💯 

We can probably merge + continue to improve. It looks like it covers all the basics for now 👍 

> I tried and a regex like m/(.+) (?:\((:.+|-webkit.+)\))/ will work

You can probably drop the outer `(?: )` -- that looks fine though 👍

**Comment by moollaza** [comment]
@rasikapohankar Thanks! This LGTM :+1:

**Comment by moollaza** [comment]
This is live on beta: 

- https://beta.duckduckgo.com/?q=webkit+overflow+scrolling&t=hh&ia=about
- https://beta.duckduckgo.com/?q=-moz-binding&t=hh&ia=iatesting
- https://beta.duckduckgo.com/?q=-webkit-autofill&t=hh&ia=iatesting

There is a triggering issue though where exact title matches don't trigger the Fathead as expected. This s because of the leading `-` and we'll need to fix this internally.

**Comment by rasikapohankar** [comment]
Thanks @moollaza @hchienjo!

**Comment by moollaza** [comment]
@rasikapohankar @hchienjo this is now live!

https://duckduckgo.com/?q=css+moz+loading&ia=about

Great work 👍  💯

**Pull request merged**